### PR TITLE
3D Keyboard 

### PR DESCRIPTION
**Pull request by paseb**
This is the added 3D keyboard for text entry.  The keyboard and KeyboardInputField support different styles of keyboard to be targeted.  Currently there is a bug with getting HoloLens Input Module to not create double input (https://forums.hololens.com/discussion/6999/ui-button-on-click-event-firing-twice).

### CODE REVIEW

File: ProjectSettings/GraphicsSettings.asset [hunk 1]
```diff
@@ -39,9 +39,21 @@ GraphicsSettings:
   m_PreloadedShaders: []
   m_SpritesDefaultMaterial: {fileID: 10754, guid: 0000000000000000f000000000000000,
     type: 0}
+<<<<<<< HEAD
```
Comment (StephenHodgson) [inline ProjectSettings/GraphicsSettings.asset]: Looks broken

File: ProjectSettings/ProjectSettings.asset [hunk 2]
```diff
@@ -210,7 +210,10 @@ PlayerSettings:
   iOSMetalForceHardShadows: 0
   metalEditorSupport: 1
   metalAPIValidation: 1
+<<<<<<< HEAD
+=======
   iOSRenderExtraFrameOnPause: 1
+>>>>>>> ea6c94daea2cfe3366daef3abbd09d429b0f439e
```
Comment (StephenHodgson) [inline ProjectSettings/ProjectSettings.asset]: Same here.

File: ProjectSettings/ProjectVersion.txt [hunk 3]
```diff
@@ -1 +1,9 @@
+<<<<<<< HEAD
+<<<<<<< HEAD
+m_EditorVersion: 5.6.0f3
+=======
+m_EditorVersion: 5.6.1f1
+>>>>>>> ea6c94daea2cfe3366daef3abbd09d429b0f439e
+=======
 m_EditorVersion: 5.6.2f1
+>>>>>>> e884de114b8d267c45b534bab5894651c8ff4d90
```
Comment (StephenHodgson) [inline ProjectSettings/ProjectVersion.txt]: Broken

File: ProjectSettings/GraphicsSettings.asset (hunk 1 above)
Comment (paseb) [inline ProjectSettings/GraphicsSettings.asset]: What's the best way to revert?  I thought I had but obviously not.  D'oh!

thanks,
-pat

**Comment by paseb** [comment]
Settings should all now be reverted.

File: Assets/HoloToolkit/UI/Scripts/AxisSlider.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        public float slideAccel = 5.25f;
+        public float slideFriction = 6f;
+        public float deadZone = 0.55f;
+        public float clampDistance = 300.0f;
+        public float bounce = 0.5f;
+
+        [HideInInspector]
+        public Vector3 TargetPoint;
+
+        float GetAxis(Vector3 v)
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/UI/Scripts/AxisSlider.cs]: nit: explicit access modifiers. (Is this public or private?)

File: Assets/HoloToolkit/UI/Scripts/AxisSlider.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                case EAxis.Y: v.y = f; break;
+                case EAxis.Z: v.z = f; break;
+            }
+            return v;
+        }
+
+        /// <summary>
+        /// Use late update to track the input slider
+        /// </summary>
+        public void LateUpdate()
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/UI/Scripts/AxisSlider.cs]: nit: All the Unity API methods should be private, not public.

File: Assets/HoloToolkit/UI/Scripts/CapsLockHighlight.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+		/// </summary>
+		[SerializeField]
+		private Image m_Highlight;
+
+		/// <summary>
+		/// The keyboard to check for caps locks
+		/// </summary>
+		private Keyboard m_Keyboard;
+
+
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/UI/Scripts/CapsLockHighlight.cs]: nit: extra line space

File: Assets/HoloToolkit/UI/Scripts/CapsLockHighlight.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+		private Keyboard m_Keyboard;
+
+		
+		/// <summary>
+		/// Unity Start method.
+		/// </summary>
+		private void Start()
+		{
+			m_Keyboard = this.GetComponentInParent<Keyboard>();
+			this.UpdateState();
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/UI/Scripts/CapsLockHighlight.cs]: nit: this. is implied, we can safely remove it.

File: Assets/HoloToolkit/UI/Scripts/KeyboardInputField.cs [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        {
+            // Unsubscribe from delegate functions
+            Keyboard.Instance.onTextUpdated -= this.Keyboard_onTextUpdated;
+            Keyboard.Instance.onClosed -= this.Keyboard_onClosed;
+        }
+    }
+
+#if UNITY_EDITOR
+    [CustomEditor(typeof(KeyboardInputField))]
+    public class KeyboardInputFieldEditor : Editor
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/UI/Scripts/KeyboardInputField.cs]: This should be broken out into it's own class, in an Editor folder.

File: Assets/HoloToolkit/UI/Scripts/SliderInputField.cs [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+        /// <summary>
+        /// Override OnDeselect
+        /// </summary>
+        /// <param name="eventData"></param>
+        public override void OnDeselect(BaseEventData eventData)
+        {
+            // Do nothing for deselection
+        }
+    }
+
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/UI/Scripts/SliderInputField.cs]: nit: extra line space

File: Assets/HoloToolkit/UI/Scripts/SymbolDisableHighlight.cs [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        /// The text field to update.
+        /// </summary>
+        [SerializeField]
+        private Image m_ImageField;
+
+        /// <summary>
+        /// The color to switch to when the button is disabled.
+        /// </summary>
+        [SerializeField]
+		private Color m_DisabledColor = Color.grey;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/UI/Scripts/SymbolDisableHighlight.cs]: nit: tab formatting

File: Assets/HoloToolkit/UI/Scripts/SymbolDisableHighlight.cs [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+            if (m_ImageField != null)
+            {
+                m_StartingColor = m_ImageField.color;
+            }
+
+            m_Button = this.GetComponentInParent<Button>();
+
+			this.UpdateState();
+
+
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/UI/Scripts/SymbolDisableHighlight.cs]: nit: extra spaces

File: Assets/HoloToolkit/UI/Scripts/UICollection.cs [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+    /// AddItem(RectTransform) method. UI elements that are added to this
+    /// collection via the Editor will be automatically arranged when this
+    /// component executes.
+    /// To use this component attach it to a UI element (a GameObject with a
+    /// RectTransform component) such as an Image or Panel.
+    /// </summary>
+    [AddComponentMenu("UI/HoloUIKit/UICollection")]
+    [RequireComponent(typeof(RectTransform))]
+    [ExecuteInEditMode]
+    public class UICollection : MonoBehaviour
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/UI/Scripts/UICollection.cs]: nit: tab formatting

File: Assets/HoloToolkit/UI/Scripts/UICollection.cs [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+        /// </summary>
+        public float VerticalSpacing = 0.0f;
+
+        /// <summary>
+        /// A list of items in this collection. This list should not be modified directly. Instead
+        /// use AddItem(RectTransform) and RemoveItem(RectTransform).
+        /// </summary>
+        public List<RectTransform> Items { get; private set; }
+
+        void Awake()
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/UI/Scripts/UICollection.cs]: nit: explicit private access modifier for Unity APIs

File: Assets/HoloToolkit/UI/Scripts/UICollection.cs [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+                // Set to the max height
+                updatedSize.y = MaxHeight;
+            }
+
+            Vector2 currentOffset = Vector2.zero;
+
+            float columnHeight = 0.0f;
+            float maxPanelWidth = 0.0f;
+
+            foreach (RectTransform item in Items)
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/UI/Scripts/UICollection.cs]: We could switch to a for loop seeing as this is in an update loop.

File: Assets/HoloToolkit/UI/Scripts/UICollection.cs [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+            {
+                // Ensure the anchors and pivot are set properly for positioning in the UICollection
+                item.anchorMin = new Vector2(0.0f, 1.0f);
+                item.anchorMax = new Vector2(0.0f, 1.0f);
+                item.pivot = new Vector2(0.0f, 1.0f);
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/UI/Scripts/UICollection.cs]: We could cache this vector allocation as well bc of the loop.

File: Assets/HoloToolkit/UI/Scripts/KeyboardValueKey.cs [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+        /// </summary>
+        private void Awake()
+        {
+            m_Button = GetComponent<Button>();
+        }
+
+        /// <summary>
+        /// Initialize key text, subscribe to the onClick event, and subscribe to keyboard shift event.
+        /// </summary>
+        void Start()
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/UI/Scripts/KeyboardValueKey.cs]: nit: private

File: Assets/HoloToolkit/UI/Scripts/UICollection.cs [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+        /// </summary>
+        /// <param name="item">The UI element to add to the collection.</param>
+        public void AddItem(RectTransform item)
+        {
+            _AddItem(item);
+
+            UpdateLayout();
+        }
+
+        private void _AddItem(RectTransform item)
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/UI/Scripts/UICollection.cs]: Why does this method have an underscore prefix?

File: Assets/HoloToolkit/UI/Scripts/UICollection.cs [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+                if (childRect != null)
+                {
+                    _AddItem(childRect);
+                }                
+            }
+        }
+
+        protected virtual void UpdateLayout()
+        {
+            RectTransform rectTransform = GetComponent<RectTransform>();
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/UI/Scripts/UICollection.cs]: Nit: we should cache this component in awake or start.

File: Assets/HoloToolkit/UI/Scripts/SymbolDisableHighlight.cs [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+        /// The text field to update.
+        /// </summary>
+        [SerializeField]
+        private Image m_ImageField;
+
+        /// <summary>
+        /// The color to switch to when the button is disabled.
+        /// </summary>
+        [SerializeField]
+        private Color m_DisabledColor = Color.grey;
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/UI/Scripts/SymbolDisableHighlight.cs]: nit: tab formatting

File: Assets/HoloToolkit/UI/Scripts/UICollection.cs [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+            {
+                Debug.LogError("This component must be attached to a GameObject with a RectTransform component!");
+            }
+
+            // Collect children items already added (likely added in the Editor)
+            CollectItems();
+            UpdateLayout();
+        }
+
+        void Update()
```
Comment (StephenHodgson) [inline Assets/HoloToolkit/UI/Scripts/UICollection.cs]: nit: private

**Comment by paseb** [comment]
Hi @StephenHodgson, 

I should have addressed all our comments with what's currently committed.  Please let me know if you see anything else.  Also let me know if you have any input on the keyboard itself if you've had the chance to test it.

thanks,
-pat

**Comment by StephenHodgson** [comment]
Hey Pat, sorry I won't be able to have the chance to test until later today.

**Comment by paseb** [comment]
No problem, I just wanted to make sure I didn't miss anything :).  Thanks Stephen!

File: Assets/HoloToolkit/UI/Scripts/UICollection.cs [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+        /// layout to update immediately.
+        /// NOTE: The added item's RectTransform will get modified in order to layout properly in this collection.
+        /// </summary>
+        /// <param name="item">The UI element to add to the collection.</param>
+        public void AddItem(RectTransform item)
+        {
+            Items.Add(item);
+
+            item.SetParent(transform);
+            item.transform.localScale = new Vector3(1.0f, 1.0f, 1.0f);
```
Comment (NeerajW) [inline Assets/HoloToolkit/UI/Scripts/UICollection.cs]: NIT: Vector3.one

**Comment by cre8ivepark** [comment]
Awesome! Once it is merged, I'll do visual fit & finish pass and look into the possibility of adding Japanese Keyboard too. (https://github.com/Microsoft/HoloToolkit-Unity/issues/735)

**Comment by NeerajW** [comment]
@StephenHodgson will wait on you to approve and then merge.

**Pull request merged**